### PR TITLE
refactor: Migrate network pkg to Azure SDK v2 - Part 3

### PR DESCRIPTION
**Pull request by helayoty**
We mostly follow this guide for code migration:
https://github.com/Azure/azure-sdk-for-go/blob/main/documentation/MIGRATION_GUIDE.md

Signed-off-by: Heba Elayoty <hebaelayoty@gmail.com>

### CODE REVIEW

File: pkg/network/aci_network.go [hunk 1]
```diff
@@ -101,33 +111,60 @@ func (pn *ProviderNetwork) SetVNETConfig(ctx context.Context, azConfig *auth.Con
...
-			return fmt.Errorf("unable to delegate subnet '%s' to Azure Container Instance since it references the route table '%s'", pn.SubnetName, *subnet.SubnetPropertiesFormat.RouteTable.ID)
+		if currentSubnet.Properties.RouteTable != nil {
+			return fmt.Errorf("unable to delegate subnet '%s' to Azure Container Instance since it references the route table '%s'", pn.SubnetName, *currentSubnet.Properties.RouteTable.ID)
 		}
-		if subnet.SubnetPropertiesFormat.ServiceAssociationLinks != nil {
-			for _, l := range *subnet.SubnetPropertiesFormat.ServiceAssociationLinks {
-				if l.ServiceAssociationLinkPropertiesFormat != nil {
-					if *l.ServiceAssociationLinkPropertiesFormat.LinkedResourceType == subnetDelegationService {
+		if currentSubnet.Properties.ServiceAssociationLinks != nil {
+			for _, l := range currentSubnet.Properties.ServiceAssociationLinks {
```
Comment (Fei-Guo) [inline pkg/network/aci_network.go:165]: Do you miss a * before currentSubnet?

File: pkg/network/aci_network.go [hunk 2]
```diff
@@ -101,33 +111,60 @@ func (pn *ProviderNetwork) SetVNETConfig(ctx context.Context, azConfig *auth.Con
...
+	response, err := subnetsClient.Get(ctx, pn.VnetResourceGroup, pn.VnetName, pn.SubnetName, nil)
 	if err != nil && !errdefs.IsNotFound(err) {
 		return fmt.Errorf("error while looking up subnet: %v", err)
 	}
 	if errdefs.IsNotFound(err) && pn.SubnetCIDR == "" {
 		return fmt.Errorf("subnet '%s' is not found in vnet '%s' in resource group '%s' and subscription '%s' and subnet CIDR is not specified", pn.SubnetName, pn.VnetName, pn.VnetResourceGroup, pn.VnetSubscriptionID)
 	}
 	if err == nil {
+		currentSubnet := response.Subnet
 		if pn.SubnetCIDR == "" {
```
Comment (Fei-Guo) [inline pkg/network/aci_network.go:155]: I suggest to add a nil check for currentSubnet.Properties.AddressPrefix

if currentSubnet.Properties.AddressPrefix != nil {

if pn.SubnetCIDR == "" {
    pn.SubnetCIDR = *currentSubnet.Properties.AddressPrefix
}

if pn.SubnetCIDR != *currentSubnet.Properties.AddressPrefix {
			   return fmt.Errorf("found subnet '%s' using different CIDR: '%s'. desired: '%s'", pn.SubnetName, 
    *currentSubnet.Properties.AddressPrefix, pn.SubnetCIDR)
		}
}

File: pkg/network/aci_network.go (hunk 1 above)
Comment (helayoty) [reply pkg/network/aci_network.go:165]: No, the new SDK changed the type to `[]*string` instead of `*[]string`.

File: pkg/network/aci_network.go [hunk 3]
```diff
@@ -100,33 +110,61 @@ func (pn *ProviderNetwork) SetVNETConfig(ctx context.Context, azConfig *auth.Con
...
+		if currentSubnet.Properties.RouteTable != nil {
+			return fmt.Errorf("unable to delegate subnet '%s' to Azure Container Instance since it references the route table '%s'", pn.SubnetName, *currentSubnet.Properties.RouteTable.ID)
 		}
-		if subnet.SubnetPropertiesFormat.ServiceAssociationLinks != nil {
-			for _, l := range *subnet.SubnetPropertiesFormat.ServiceAssociationLinks {
-				if l.ServiceAssociationLinkPropertiesFormat != nil {
-					if *l.ServiceAssociationLinkPropertiesFormat.LinkedResourceType == subnetDelegationService {
+		if currentSubnet.Properties.ServiceAssociationLinks != nil {
+			for _, l := range currentSubnet.Properties.ServiceAssociationLinks {
+				if l.Properties != nil {
```
Comment (Fei-Guo) [inline pkg/network/aci_network.go:146]: I suggest doing a null check here.

if l.Properties != nil && l.Properties.LinkedResourceType != nil {
  if *l.Properties.LinkedResourceType == subnetDelegationService {
  ...
}
}

File: pkg/network/aci_network.go [hunk 4]
```diff
@@ -135,8 +173,8 @@ func (pn *ProviderNetwork) setupNetwork(ctx context.Context, azConfig *auth.Conf
 				}
 			}
 		} else {
-			for _, d := range *subnet.SubnetPropertiesFormat.Delegations {
-				if d.ServiceDelegationPropertiesFormat != nil && *d.ServiceDelegationPropertiesFormat.ServiceName == subnetDelegationService {
+			for _, d := range currentSubnet.Properties.Delegations {
+				if d.Properties != nil && *d.Properties.ServiceName == subnetDelegationService {
```
Comment (Fei-Guo) [inline pkg/network/aci_network.go:157]: I suggest doing a null check here.

if d.Properties != nil && d.Properties.ServiceName!= nil && *d.Properties.ServiceName == subnetDelegationService {

File: pkg/provider/aci.go [hunk 5]
```diff
@@ -472,8 +474,11 @@ func (p *ACIProvider) deleteContainerGroup(ctx context.Context, podNS, podName s
 			false,
 		)
 
-		if updateErr != nil && !errdefs.IsNotFound(updateErr) {
-			log.G(ctx).WithError(updateErr).Errorf("failed to update termination status for cg %v", cgName)
```
Comment (Fei-Guo) [inline pkg/provider/aci.go:476]: Why do you change ths? updateErr is not returned by SDK call

Comment (helayoty) [reply pkg/provider/aci.go]: Good catch. Fixed.

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**